File: python/leetcode055.py
class Solution:
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        '''
        ls = len(nums)
        if nums[0] >= ls-1:
            return True
        if nums[0]==0:
            return False
        for i in range(1,nums[0]+1):
            if self.canJump(nums[i:]):
                return True
        return False
        '''
        length = len(nums)
        begin = length-1
        for i in reversed(range(length-1)):
            if i + nums[i] >= begin:
                begin = i
        return not begin

if __name__ == "__main__":
    s = Solution()
    print(s.canJump([1,2,3]))

File: leetcodec++/leetcode055.cpp
//
// Created by 黄敏辉 on 2019-07-16.
//

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int len = nums.size();
        if (len == 1)
            return true;
        int pos = 0;
        int max_dis = nums[0];

        while (max_dis < len-1){
            int flag = 0;
            for (int i = pos; i <= pos + nums[pos]; i++){
                if (i + nums[i] > max_dis){
                    pos = i;
                    max_dis = i + nums[i];
                    flag = 1;
                    break;
                }
            }
            if (!flag)
                return false;
        }
        return true;
    }
};